Convert SAM SoC SERIAL drivers to DT macros

nandojve opened the pull request:
This convert Atmel SAM/SAM0 SERIAL drivers to use DT macros.

see #23107

nandojve committed:
drivers: serial: sam0: uart: convert to DT macros

Convert driver to use new device tree macros.

see #23107

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: serial: sam: uart: convert to DT macros

Convert driver to use new device tree macros.

see #23107

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: serial: sam: usart: convert to DT macros

Convert driver to use new device tree macros.

see #23107

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

stephanosio commented on drivers/serial/uart_sam.c:
```diff
@@ -12,53 +14,51 @@
-#if CONFIG_UART_SAM_PORT_0 == 1
+#if DT_HAS_NODE(DT_NODELABEL(uart0))
 
-#if DT_UART_SAM_PORT_0_BAUD_RATE == 0
-#error "DT_UART_SAM_PORT_0_BAUD_RATE has to be bigger than 0"
+#if DT_PROP(DT_NODELABEL(uart0), current_speed) == 0
+#error "uart0 current-speed property has to be bigger than 0"
 #endif
```
Maybe these can be moved into the block that does `UART_SAM_DEVICE_INIT(n);` or even better into `UART_SAM_DEVICE_INIT` macro itself using `BUILD_ASSERT`?

nandojve replied:
Done! it was need create a macro to check if prop exists. This way we can avoid long list of errors when property not exists.

stephanosio commented on drivers/serial/uart_sam0.c:
```diff
@@ -925,21 +925,22 @@ static const struct uart_driver_api uart_sam0_driver_api = {
+#define UART(n)		DT_NODELABEL(sercom##n)
+#define HAS_UART(n)	DT_NODE_HAS_COMPAT(UART(n), atmel_sam0_uart)
 
+#if CONFIG_UART_INTERRUPT_DRIVEN || CONFIG_UART_ASYNC_API
```
This should be:
`#if defined(CONFIG_UART_INTERRUPT_DRIVEN) || defined(CONFIG_UART_ASYNC_API)`

nandojve replied:
Done!

stephanosio reviewed (commented):
I can see that we cannot use `DT_INST_n` here because the SAM/SAM0 device tree does not fully incorporate device clock and I/O mappings, and `DT_INST_n` does not necessarily match the `UART_n` or `SERCOM_n` definitions used to infer device clock and I/Os.

This should eventually be fixed, but I guess that is out of scope for this PR.

stephanosio commented on drivers/serial/uart_sam.c:
```diff
@@ -364,206 +364,74 @@ static const struct uart_driver_api uart_sam_driver_api = {
+	.periph_id = DT_PROP(DT_NODELABEL(uart##n), peripheral_id),	\
+	.pin_rx = PIN_UART##n##_RXD,					\
+	.pin_tx = PIN_UART##n##_TXD,					\
+	UART_SAM_IRQ_CALLBACK(n)					\
```
On a somewhat unrelated note, as static device tree support grows more complex and advanced, I wonder if we should start considering the introduction of an external preprocessor/codegen for this purpose (e.g. using m4, or modern equivalent).

Working with these multi-line C preprocessor macros is hard enough and using `#if/#ifdef`s inside one is impossible.

cc @galak @mbolivar-nordic

mbolivar-nordic replied:
I am not yet convinced this is necessary and I would like to avoid it if possible. Having Kconfig and DT is already a big learning curve for traditional bare metal embedded developers. Adding another meta/config language to Zephyr proper is a big step. Nothing is stopping out of tree driver writers from doing what they want, of course.

nandojve committed:
dts: arm: atmel: sam3x: Rename uart to uart0

The Atmel UART driver now uses new DT macros and need use NODELABEL to
select proper node. The driver now needs search by name and index to
find the right node. This rename SAM3X uart to uart0 to be compliant.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve commented:
rebase and add d5d1639 to fix sam3x. Now we need all nodes with index.

galak commented:
The SAM (not SAM0) part of this should now be handled by #24319

nandojve commented:
Close in favor of #24319, #24572.
